service/frontend/workflowHandler.go

```diff
@@ -965,7 +965,7 @@ func (wh *WorkflowHandler) RespondWorkflowTaskFailed(
 		request.Failure = serverFailure
 	}
 
-	if request.GetCause() == enumspb.WORKFLOW_TASK_FAILED_CAUSE_UNHANDLED_COMMAND {
+	if request.GetCause() == enumspb.WORKFLOW_TASK_FAILED_CAUSE_UNSPECIFIED {
```

Comment:
WORKFLOW_TASK_FAILED_CAUSE_UNSPECIFIED is the default value for WorkflowTaskFailedCause. Anywhere that the SDK's are currently sending back a RespondWorkflowTaskFailedRequest without explicitly setting a cause will now be interpreted as a nondeterministic error (from a metrics and logging point of view). Is this OK?

Reply:
> WORKFLOW_TASK_FAILED_CAUSE_UNSPECIFIED is the default value for WorkflowTaskFailedCause. Anywhere that the SDK's are currently sending back a RespondWorkflowTaskFailedRequest without explicitly setting a cause will now be interpreted as a nondeterministic error (from a metrics and logging point of view). Is this OK?

depending on how SDK works, requested review from SDK team

Reply:
I did a quick grep on sdk code, it looks like sdk-go only set this field to WORKFLOW_TASK_FAILED_CAUSE_WORKFLOW_WORKER_UNHANDLED_FAILURE when there is any unhandled error including non-deterministic error; and java-sdk only set this to WORKFLOW_TASK_FAILED_CAUSE_RESET_STICKY_TASK_QUEUE when it sees a task whose workflow state is already evicted from local cache. It does not seem like either SDK is properly reporting non-deterministic error clearly to server.

Reply:
It's a little tricky because, as noted above, the default value for `WorkflowTaskFailedCause` is `WORKFLOW_TASK_FAILED_CAUSE_UNSPECIFIED`. So if the sdk leaves this value unset in the `RespondWorkflowTaskFailedRequest.Cause` proto field, it will indeed trigger the "nondetermistic" error code in the server.

That is what is happening in the Java SDK here: https://github.com/temporalio/sdk-java/blob/60dbb99d24c9fd31fec220b9c5a6502f16e44dcf/temporal-sdk/src/main/java/io/temporal/internal/replay/ReplayWorkflowTaskHandler.java#L164

I've verified with manual testing that with @wxing1292's change, the nondeterministic counter gets bumped as expected when a version error occurs.

My initial testing with Go SDK indicates that is NOT happening however. So perhaps the Go SDK is setting an explicit value for `RespondWorkflowTaskFailedRequest.Cause` that is different from `WORKFLOW_TASK_FAILED_CAUSE_UNSPECIFIED`

Reply:
Shouldn't it be fixed on Java SDK side then?

Reply:
I would suggest adding a new enum specifically for non-deterministic error  (say WORKFLOW_TASK_FAILED_CAUSE_NON_DETERMINISTIC_ERROR) if we want to handle it. And change SDKs to report that when they encounter the non-deterministic error. Then server side could use that to report metrics accordingly.

Reply:
WORKFLOW_TASK_FAILED_CAUSE_UNSPECIFIED is default value and treat it as non deterministic error is not right.

Reply:
Agree with @yiminc. There should be a specific value for nondeterminism errors. The SDKs rarely set this cause value to anything meaningful. Fixing Java and Go to attach more meaningful errors should be done, but FYI it's not a small task, the error paths are often nonobvious.

Ideally, there would be separate enums for values the SDK can set, and values the server can set. As it stands they are combined in a nonobvious way. Of course, not really doable in a backwards compat way. At minimum, the protos should document what causes are settable by whom and what they mean (in general the protos are woefully undocumented).

Reply:
Grep on sdk-java shows WORKFLOW_TASK_FAILED_CAUSE_UNHANDLED_COMMAND is only used by testservice.

Reply:
Not related to this PR. But WORKFLOW_TASK_FAILED_CAUSE_UNHANDLED_COMMAND really should be named WORKFLOW_TASK_FAILED_CAUSE_UNHANDLED_EVENTS because it is set by server when workflow try to close and there are unseen events (buffered events) that workflow should take a look before closing.

Reply:
It does sound like adding a new enum value to signal nondeterminism detected from the client is the best solution. What is the backwards compatibility issue with doing this? The only place where the server is currently checking for WORKFLOW_TASK_FAILED_CAUSE_UNHANDLED_COMMAND is where it logs and emits metrics for `service_errors_nondeterministic` which is currently broken anyway.